client/state/plugins/installed/selectors.ts

```diff
@@ -0,0 +1,382 @@
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
+	( state ) => {
+		const allPluginsIndexedByPluginSlug = getAllPluginsIndexedByPluginSlug( state );
+
+		return Object.values( allPluginsIndexedByPluginSlug ).reduce(
+			(
+				pluginsIndexedBySiteId: { [ siteId: number ]: { [ pluginSlug: string ]: Plugin } },
+				plugin
+			) => {
+				Object.keys( plugin.sites )
+					.map( ( siteId ) => Number( siteId ) )
+					.forEach( ( siteId ) => {
+						pluginsIndexedBySiteId[ siteId ] = {
+							...pluginsIndexedBySiteId[ siteId ],
+							[ plugin.slug ]: plugin,
+						};
+					} );
+
+				return pluginsIndexedBySiteId;
+			},
+			{}
+		);
+	},
+	( state ) => [ getAllPluginsIndexedByPluginSlug( state ), getSiteIdsThatHavePlugins( state ) ]
+);
+
+export const getFilteredAndSortedPlugins = createSelector(
+	( state: AppState, siteIds: number[], pluginFilter?: PluginFilter ) => {
+		const allPluginsIndexedBySiteId = getAllPluginsIndexedBySiteId( state );
+
+		// Properties on the objects in allPluginsIndexedBySiteId will be modified and the
+		// selector memoization always returns the same object, so use cloneDeep to avoid
+		// altering it for everyone.
+		const allPluginsForSites: { [ pluginSlug: string ]: Plugin } = cloneDeep(
+			siteIds
+				.map( ( siteId: number ) => allPluginsIndexedBySiteId[ siteId ] )
+				.filter( ( plugin ) => !! plugin )
+				.reduce( ( accumulator, current ) => ( { ...accumulator, ...current } ), {} )
+		);
+
+		// Filter the sites object on the plugins so that only data for the requested siteIds is present
+		for ( const pluginSlug of Object.keys( allPluginsForSites ) ) {
+			allPluginsForSites[ pluginSlug ].sites = pick(
+				allPluginsForSites[ pluginSlug ].sites,
+				siteIds
+			);
+		}
+
+		const pluginList =
+			pluginFilter && _filters[ pluginFilter ]
+				? filter( allPluginsForSites, _filters[ pluginFilter ] )
+				: allPluginsForSites;
+
+		return sortBy( pluginList, ( plugin: Plugin ) => plugin.slug.toLowerCase() ) as Plugin[];
+	},
+	( state: AppState ) => [ getAllPluginsIndexedBySiteId( state ) ],
+	( state: AppState, siteIds: number[], pluginFilter?: PluginFilter ) => {
+		return [ siteIds, pluginFilter ].flat().join( '-' );
+	}
+);
+
+export function getPluginsWithUpdates( state: AppState, siteIds: number[] ) {
+	return filter( getFilteredAndSortedPlugins( state, siteIds, undefined ), _filters.updates ).map(
+		( plugin ) => ( {
+			...plugin,
+			type: 'plugin',
+		} )
+	);
+}
+
+export function getPluginsOnSites( state: AppState, plugins: Plugin[] ) {
+	return Object.values( plugins ).reduce(
+		( acc: { [ pluginSlug: string ]: Plugin }, plugin: Plugin ) => {
+			const siteIds = Object.keys( plugin.sites ).map( ( x ) => Number( x ) );
```

Comment:
I could, but I wouldn't rely on future developers to know about and use such a function when it is so obvious that the sites object is indexed by site ID. There was lots of mismatching between numbers and strings before because not all code was written to handle object keys always being strings, but TypeScript has fixed all of that so such mistakes shouldn't be possible going forward.